[EMS] Create dev tooling for config plugins (#11574)
expo/expo: packages/expo-module-scripts/README.md

You should also add the following fields to your `package.json`:

### 🔌 Config plugin

To create a [config plugin](https://github.com/expo/expo-cli/blob/master/packages/config-plugins/README.md) that automatically configures your native code, you have two options:

1. Create a `plugin` folder and write your plugin in TypeScript (Recommended).

> 2. Create an `app.plugin.js` and write the plugin in pure Node.js compliant JavaScript

Reviewer: ```suggestion
2. Create an `app.plugin.js` in the project root and write the plugin in pure Node.js compliant JavaScript
```

> Config plugins must be transpiled for Node.js (LTS) which is slightly different to native React modules which support features like the import/export keywords. This means we'll need two different `tsconfig.json` files and two different `src` (and `build`) folders.

Reviewer: ```suggestion
Config plugins must be transpiled for compatibility with Node.js (LTS). The features supported in Node.js are slightly different to React Native modules, which support ES6 import/export keywords and JSX, for example. This means we'll need two different `tsconfig.json` files and two different `src` (and `build`) folders - one for the code that will execute in a React Native app and the other for the plugin that executes in Node.js.
```

```

Create a `app.config.js` (the entry point for a config plugin):

```js
module.exports = require('./plugin/build');
```

Create a `plugin/tsconfig.json` file. Notice that this uses `tsconfig.plugin` as the base config:

```json
{
  "extends": "expo-module-scripts/tsconfig.plugin",
  "compilerOptions": {
    "outDir": "build",
    "rootDir": "src"
  },
  "include": ["./src"],
  "exclude": ["**/__mocks__/*", "**/__tests__/*"]
}
```

In your `plugin/src/index.ts` file, write your TypeScript config plugin:

```ts
import { ConfigPlugin } from '@expo/config-plugins';

const withNewName: ConfigPlugin<{ name?: string }> = (config, { name = 'my-app' } = {}) => {
  config.name = name;
  return config;
};

export default withNewName;
```

> 💡 Tip: Using named functions makes debugging easier with `EXPO_DEBUG=true`

Optionally, you can add `plugin/jest.config.js` to override the default project Jest preset.

```ts
module.exports = require('expo-module-scripts/jest-preset-plugin');
```

Use the following scripts to interact with the plugin:

- `yarn build plugin`: Build the plugin.
- `yarn clean plugin`: Delete the `plugin/build` folder.
- `yarn lint plugin`: Lint the `plugin/src` folder.

> - `yarn test plugin`: Alias for `npx jest --rootDir ./plugin --config ./plugin/jest.config.js`, uses the project jest preset if `plugin/jest.config.js` doesn't exist.

Reviewer 2: ```suggestion
- `yarn test plugin`: Alias for `npx jest --rootDir ./plugin --config ./plugin/jest.config.js`, uses the project's Jest preset if `plugin/jest.config.js` doesn't exist.
```
